src/util/tipRadiusHelpers.ts

```diff
@@ -44,13 +47,16 @@ const determineLocationMatch = (node, geoResolution, geoValueToMatch) => {
 * @param tree
 * @returns null (if data not ready) or array of tip radii
 */
-export const calcTipRadii = ({tipSelectedIdx = false, selectedLegendItem = false, geoFilter = [], searchNodes = false, colorScale, tree}) => {
+export const calcTipRadii = (
+  {tipSelectedIdx = false, selectedLegendItem = false, geoFilter = [], searchNodes = false, colorScale, tree}:
+  {tipSelectedIdx:(number|false), selectedLegendItem:(number|string|false), geoFilter:([string, string]|[]), searchNodes:any, colorScale:any, tree:any}
+) => {
   if (selectedLegendItem !== false && tree && tree.nodes) {
-    return tree.nodes.map((d) => determineLegendMatch(selectedLegendItem, d, colorScale) ? tipRadiusOnLegendMatch : tipRadius);
+    return tree.nodes.map((d:any) => determineLegendMatch(selectedLegendItem, d, colorScale) ? tipRadiusOnLegendMatch : tipRadius);
   } else if (geoFilter.length===2 && tree && tree.nodes) {
-    return tree.nodes.map((d) => determineLocationMatch(d, ...geoFilter) ? tipRadiusOnLegendMatch : tipRadius);
```

Comment:
I was surprised TS didn't allow `...geoFilter` given that the preceding `if` statement ensures there are 2 elements. Perhaps three's some gotchas in the spread operator that I don't understand.

Reply:
Compiled languages rarely check array lengths[1], because it often requires runtime information, while compilers only have information available at compile time. In this particular case, it might be possible to prove that array size is 2, but in many cases it's not.

I'd say, this code is questionable. If `geoFilter` always has two items, then why it's not an object with named fields. They are even given names inside `determineLocationMatch()`. 

It looks like an abuse of container type. I guess it can also be empty (parameter is defaulted to `[]`)? Can it be anything else? 1 element? How about 3? Long story short, is'a a bit of a sloppy, indecisive code, because it does not know what it wants. 

Typescript does not make it easy to write code where things constantly transmute - an elephant becomes a banana and then in another file it becomes a space ship. Typescript requires some discipline. It puts constraints onto the Wild West of JS, and tries to make code less dynamic and easier to reason about.


Ideally I'd write something like:

```typescript
// TODO: prehaps rename fields to have less of repeated "geo" prefix
export interface GeoFilters {
  geoResolution: string;
  geoValueToMatch: string;
}

function calcTipRadii(
  // ... more params
  geoFilters?: GeoFilters // can be `undefined`
  // ... more params
) {
  // ...
  if (geoFilters) {
    return determineLocationMatch(node, geoFilters)
  }
  // ...
}

function determineLocationMatch(
  node: Node,
  { geoResolution, geoValueToMatch }: GeoFilters // destructuring
) {
  return geoValueToMatch === getTraitFromNode(node, geoResolution);
}

```


If you really want to abuse the container type (e.g. because it comes from a stable file format, and you don't want to write additional conversions), then in Typescript there's a [tuple](https://www.typescriptlang.org/docs/handbook/2/objects.html#tuple-types) type, which can emulate fixed-size array. You could make `geoTypes` of type `[string, string]`, which is different from `string[]`. But in this particular code `geoTypes` defaults to `[]`, and `[]` cannot be of type `[string, string]`. So it needs additional re-thinking.


--

[1] One exception is Rust, where you have separate [`Vec`](https://doc.rust-lang.org/std/vec/struct.Vec.html) type for dynamic arrays and [array](https://doc.rust-lang.org/std/primitive.array.html) type `[i32;N]` for fixed-size arrays. The `N` in the array type is then a part of the type, and is checked on compile-time, while `Vec` is only checked at runtime.

Reply:
Thanks @ivan-aksamentov - since TS [allows conditional statements to narrow the types](https://www.typescriptlang.org/docs/handbook/2/narrowing.html) I expected this to be ok; perhaps it doesn't like it because we may have overriden the `length` prototype?

And fully agree with your general comments about the code -- this is one of the reasons I want to mostly write things in TS going forward (as well as the implicit documentation it'll provide when reading/writing code). However I didn't want to labour this PR with code changes, I just tried to find the most self-contained file to port to TS I could.

Reply:
Interestingly, changing the type from `geoFilter:([string, string]|[])` to `geoFilter?:[string, string]` (and therefore using `undefined` for the default argument) _does_ allow the spread operator to be used. I'm not going to make this change in this PR as there's other files which call this function using an empty array argument, but it was an interesting insight into how TS works. 

```diff
 export const calcTipRadii = (
-  {tipSelectedIdx = false, selectedLegendItem = false, geoFilter = [], searchNodes = false, colorScale, tree}:
-  {tipSelectedIdx:(number|false), selectedLegendItem:(number|string|false), geoFilter:([string, string]|[]), searchNodes:any, colorScale:any, tree:any}
+  {tipSelectedIdx = false, selectedLegendItem = false, geoFilter = undefined, searchNodes = false, colorScale, tree}:
+  {tipSelectedIdx:(number|false), selectedLegendItem:(number|string|false), geoFilter?:[string, string], searchNodes:any, colorScale:any, tree:any}
 ) => {
   if (selectedLegendItem !== false && tree && tree.nodes) {
     return tree.nodes.map((d:any) => determineLegendMatch(selectedLegendItem, d, colorScale) ? tipRadiusOnLegendMatch : tipRadius);
-  } else if (geoFilter.length===2 && tree && tree.nodes) {
-    return tree.nodes.map((d:any) => determineLocationMatch(d, geoFilter[0], geoFilter[1]) ? tipRadiusOnLegendMatch : tipRadius);
+  } else if (geoFilter?.length===2 && tree && tree.nodes) {
+    return tree.nodes.map((d:any) => determineLocationMatch(d, ...geoFilter) ? tipRadi
usOnLegendMatch : tipRadius);
   } else if (searchNodes) {
     return tree.nodes.map((d:any) => d.name.toLowerCase().includes(searchNodes) ? tipR
adiusOnLegendMatch : tipRadius);
   } else if (tipSelectedIdx) {
```